smac/epm/base_epm.py

```diff
@@ -12,64 +15,183 @@
 class AbstractEPM(object):
     '''Abstract implementation of the EPM API '''
 
-    def __init__(self, rng):
+    def __init__(self, 
+                 instance_features:np.ndarray=None,
+                 pca_dims:float=None):
         '''
         initialize random number generator
 
         Parameters
         ----------
-        rng : np.random.RandomState
+        instance_features: np.ndarray (I, K)
+            Contains the K dimensional instance features
+            of the I different instances
+        pca_dims: float
+            if set to a float, use PCA to reduce dimensionality of instance features
+            also requires to set n_feats (> pca_dims)
+        
         '''
-        self.rng = rng
+        self.instance_features = instance_features
+        self.pca_dims = pca_dims
+        if instance_features is not None:
+            self.n_feats = instance_features.shape[1]
+        else:
+            self.n_feats = 0
+        
+        self.n_params = None # will be updated on train()
+        
+        self.pca = None
+        self.scaler = None
+        if self.pca_dims and self.n_feats > self.pca_dims:
+            self.pca = PCA(n_components=self.pca_dims)
+            self.scaler = MinMaxScaler()
+            
+        # Never use a lower variance than this
+        self.var_threshold = 10 ** -5
 
     def train(self, X, Y, **kwargs):
-        '''
-        Trains the EPM on X and Y.
+        '''Trains the random forest on X and y.
 
         Parameters
         ----------
-        X: np.ndarray (N, D)
-            Input data points. The dimensionality of X is (N, D),
-            with N as the number of points and D is the number of features.
-        Y: np.ndarray (N, 1)
-            The corresponding target values.
+        X : np.ndarray [n_samples, n_features (config + instance features)]
+            Input data points.
+        Y : np.ndarray [n_samples, n_objectives]
+            The corresponding target values. n_objectives must match the
+            number of target names specified in the constructor.
```

Comment:
That's true for the `predict_marginalized_over_instances`, but for all other functions the subclass specifies the number of objectives it can handle. I could add a corresponding note in the docstrings if you wish